Add a new WarpPlateRecipe type for configurable shard recipes

edralzar opened the pull request:
This new type only applies to attunement items that can be created within the Warp Plate UI, not extra items that are mapped to a teleport effect.

The recipe result count can produce a stack larger than the item's `maxStackSize()` but the player will still be limited by it when transferring the items in their inventory.

Closes gh-410

edralzar committed:
Add a new WarpPlateRecipe type for configurable shard recipes

This new type only applies to attunement items that can be created
within the Warp Plate UI, not extra items that are mapped to a
teleport effect.

If the recipe result count field has a value bigger than the item's
maxStackSize(), then the later is used instead.

The recipe syntax is like:

```
{
  "type": "waystones:warp_plate",
  "result": {
    "item": "waystones:attuned_shard",
    "count": 1
  },
  "inner": {
    "item": "minecraft:flint"
  },
  "outer": {
    "item": "waystones:warp_dust"
  }
}
```

Closes gh-410

BlayTheNinth commented:
Added to Balm https://github.com/TwelveIterationMods/Balm/commit/a64c3bccd5bcb3b4d65d536d7a4aeb175437f420

edralzar committed:
Use Balm.getRecipes for Fabric/Forge compat

edralzar committed:
Let recipes produce more than the item's maxStackSize

If a server owner wants that, why not? The items will still not
stack in the player's inventory anyway.

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/block/entity/WarpPlateBlockEntity.java:
```diff
@@ -198,19 +200,31 @@ private boolean isEntityOnWarpPlate(Entity entity) {
-                completedFirstAttunement = true;
             }
+
+            if (recipe.getResultItem().getItem() == ModItems.attunedShard) {
```
I think this check is not necessary because the first attunement will always happen first (since it handles locking the slots during the initial opening animation) and it doesn't harm "re"-setting it to true every time.

edralzar replied:
I've removed the condition, but moved the `this.completedFirstAttunement = true` one block up so that it is only marked when a shard is created.

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/recipe/WarpPlateRecipeSerializer.java:
```diff
@@ -0,0 +1,41 @@
+                    CraftingRecipeCodecs.ITEMSTACK_OBJECT_CODEC.fieldOf("result").forGetter(WarpPlateRecipe::getResultItem),
+                    Ingredient.CODEC.fieldOf("inner").forGetter(WarpPlateRecipe::getInnerIngredient),
+                    Ingredient.CODEC.fieldOf("outer").forGetter(WarpPlateRecipe::getOuterIngredient))
+            //note that the count from the recipe is capped to the item's maxStackSize (see WarpPlateRecipe constructor)
```
comment appears to be outdated

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/recipe/WarpPlateRecipe.java:
```diff
@@ -0,0 +1,121 @@
+
+public class WarpPlateRecipe implements Recipe<WarpPlateBlockEntity> {
+    private final Ingredient innerIngredient;
+    private final Ingredient outerIngredient;
```
instead of having just a single outer ingredient, it would be good if the outer parts would be treated like a shapeless recipe, i.e. taking 4 individual (but shapedless) ingredients, so that recipes like flint + 2 glowstone + 2 redstone would also be supported

edralzar replied:
I added the option. The class now has a `List<Ingredient> getOuterShapelessIngredients()` instead and the JSON can either specify `outer` (single ingredient, repeated 4 times) or `outers` (array of 4 explicit ingredients, shapeless)

BlayTheNinth replied:
Supporting both "outer" and "outers" format adds a lot of unnecessary complexity. Since the benefit gained by having the single-ingredient "outer" variant is miniscule, especially considering how rarely this type of recipe will be defined, we should keep things simple and have just the one explicit variant of defining them.

edralzar replied:
done, and that allowed me to find a better way to validate the recipe than throwing a (server-crashing) exception 😅

BlayTheNinth reviewed (changes requested):
`initializeInventory` in `WarpPlateBlockEntity` would also have to be updated so it prefills the slots with whatever the attuned_shard.json recipe defines, since datapacks might be overriding the recipe

edralzar committed:
Add `outers` array to recipe format as an alternate shapeless list

Either use `outer` (with a single defined ingredient, implicitly
repeated 4 times) or `outers` (with all 4 ingredients explicitly
defined).

The initialization of the plate inventory also picks up the first
attuned shard recipe known rather than static ingredients.

WarpPlateRecipe has a descriptive (cached) toString().

Added an example recipe with `outers`: crush a shard with an anvil
and some dirt to get back a stack of 4 warp dust.

edralzar committed:
Mark completedFirstAttunement as soon as any shard is created

edralzar committed:
(remove unnecessary import change)

edralzar committed:
Simplify json syntax (only 'outers'), improve codec error management

Recipe requires the full `outers` array of 4 Ingredients.

The codec now makes use of `DataResult` to convey recipe syntax
errors, which doesn't prevent the server from loading in such a
case (unlike the earlier throwing of IllegalStateException in apply).

BlayTheNinth committed:
Rename visually dependant names "inner, outers" to more semantic names "primary, secondary"
Move serializer class into recipe class to match Mojang style
Avoid crashing if recipe is invalid, log error instead
Simplify initial attunement item population
Remove WarpPlateRecipe.findFirstMatchingRecipe in favor of inlining
Use more generic Recipe<Container> over Recipe<WarpPlateBlockEntity> to reduce dependency on implementation details
Merge resultItem and resultCount into one resultItem parameter in recipe ctor
Remove size check in recipe ctor since that's the codec's job
Remove emptiness check premature optimization
Remove isSameOuterIngredients premature optimization
Match codec code closer to Mojang style for better porting maintainability
Remove test recipe
